README.md
# Lamports-OTP

This server is made to work with the lamports-otp client found in the [lamports-otp-client repository](https://github.com/picturestone/lamports-otp-client).

The default user `root` with the password `root` is already available. The password is valid for 1000 logins. After that, the password must be changed. It can also be changed before. New passwords are always valid for howerver `PASSWORD_USES` is configured.

## Routes

### POST `/users`

Adds a new user. Send the following data:

```json
{
    'username': <username>,
    'password': <password>
}
```

`<username>` is the username as a string.

`<password>` is the password as a string, hashed with sha256 converted to HEX for as many times as a request to GET `auth/index` returns.

### PUT `/users`

Use this route to change the password for the logged in user. Send the following data:

```json
{
    'password': <password>
}
```

`<password>` is the password as a string, hashed with sha256 converted to HEX for as many times as a request to GET `auth/index` returns.

### GET `auth/index`

Returns the number of uses per newly set password (which is the configured parameter `PASSWORD_USES`). Returns the following data:

```json
{
    'index': <index>
}
```

`<index>` is the number of uses for which a newly set password is valid.

### POST `auth/index`

Use this route to get the number of hashes necessary for the next login. Send the following data:

```json
{
    'username': <username>
}
```

`<username>` is the username as a string.

Returns the following data:

```json
{
    'index': <index>
}
```

`<index>` is the number of hashes necessary for the password to be valid on the next login.

### POST `auth/login`

Use this route for the login. Send the following data:


```json
{
    'username': <username>,
    'password': <password>
}
```


`<username>` is the username as a string.

`<password>` is the password as a string, hashed with sha256 converted to HEX for as many times as a request to POST `auth/index` returns.

### GET `auth/logout`

Use this route to logout and invalidate the session of the currently logged in user.

## Installation

### Prerequisits

The following are prerequisits for running this website.

* git, downloadable [here](https://git-scm.com/downloads)
* node, downloadable [here](https://nodejs.org/en/download)

### Setup

This is the setup procedure:

1. Enter in the terminal: `git clone https://github.com/picturestone/lamports-otp`
1. Go to the folder `lamports-otp`
1. Download dependencies: `npm install`
1. Configurate (see Installation > Configuration) - This is optional.
1. Start the sever: `node app.js`

If all configurations are unchanged the server is reachable via the url [http://localhost:3000/](http://localhost:3000/)

### Configuration

The following parameters are configurable in the `lamports-otp/.env` file.

* `SESSION_SECRET` used to hash the session with HMAC
* `PASSWORD_USES` defines the index initially set for new accounts or changed passwords
* `PORT` is the port used by the server
* `HOST` is the host the server is running on


CHANGES.md
#Changes

#### Unreleased

- Added configuration via .env file #6
- Added change password functionality #5
- Added logout route #4
- Added login and register routes #2, #3
- Added basic npm files, added express as a setup, added hello world functionality #1

app.js
const express = require('express');
const session = require('express-session');
var cookieParser = require('cookie-parser');
let config = require('./models/configuration');
config = new config();
const authRouter = require('./routes/auth');
const userRouter = require('./routes/user');

const app = express();
app.use(cookieParser());
app.use(session({
    secret: config.sessionSecret,
    resave: true,
    saveUninitialized: true,
    cookie: {
        secure: false,
        expires: false,
        httpOnly: false
    }
}));
app.use(express.json());
// Cors settings.
app.use(function (req, res, next) {
    res.header('Access-Control-Allow-Credentials', true);
    res.header('Access-Control-Allow-Origin', req.headers.origin);
    res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE');
    res.header('Access-Control-Allow-Headers', 'X-Requested-With, X-HTTP-Method-Override, Content-Type, Accept');
    next();
});

app.use('/auth', authRouter);
app.use('/users', userRouter);

// catch 404 and forward to error handler
app.use(function (req, res, next) {
    res.sendStatus(404);
});

const port = config.port;
const host = config.host;
app.listen(port, host, () => {
    console.log(`Example app listening at http://${host}:${port}`);
})

routes/user.js
const express = require('express');
const router = express.Router();
const user = require('../models/user');
const userdao = require('../database/userdao');
let config = require('../models/configuration');
config = new config();

router.post('/', function (req, res, next) {
    username = req.body.username;
    password = req.body.password;

    if (!username || !password) {
        res.statusMessage = 'Username or password missing';
        res.sendStatus(400);
    } else {
        userdao.getByUsername(username, (foundUser) => {
            if (foundUser) {
                res.statusMessage = 'User with this username already exists';
                res.sendStatus(400);
            } else {
                userdao.insert(new user(username, password, config.index), () => {
                    res.sendStatus(201);
                });
            }
        });
    }
});

router.put('/', function (req, res, next) {
    if (!req.session.loggedin) {
        res.statusMessage = 'Not logged in';
        res.sendStatus(401);
    } else {
        password = req.body.password;

        if (!password) {
            res.statusMessage = 'Password missing';
            res.sendStatus(400);
        } else {
            userdao.getByUsername(req.session.username, (foundUser) => {
                if (foundUser) {
                    foundUser.password = password;
                    foundUser.index = config.index;
                    userdao.updateByUsername(foundUser.username, foundUser, () => {
                        res.statusMessage = 'Password changed';
                        res.sendStatus(200);
                    });
                } else {
                    res.statusMessage = 'User in session not found. Try logging in again.';
                    res.sendStatus(404);
                }
            });
        }
    }
});

module.exports = router;

database/userdao.js
const databasewrapper = require('./databasewrapper');
const user = require('../models/user');

module.exports = class userrepository {
    static getByUsername(username, cb) {
        const db = databasewrapper.getDb();
        let sql = `SELECT * FROM users WHERE username = ?`;

        db.get(sql, [username], (err, row) => {
            if(err) {
                return console.log(err.message);
            }
            if (row) {
                cb(new user(row.username, row.password, row.indexcounter));
            } else {
                cb(undefined);
            }
        });

        db.close();
    }

    static insert(user, cb) {
        const db = databasewrapper.getDb();
        let sql = `INSERT INTO users(username, password, indexcounter) VALUES(?, ?, ?)`;

        db.run(sql, [user.username, user.password, user.index], (err, row) => {
            if(err) {
                return console.log(err.message);
            }
            cb(row);
        });
    }

    static updateByUsername(username, user, cb) {
        const db = databasewrapper.getDb();
        let sql = `UPDATE users SET username = ?, password = ?, indexcounter = ? WHERE username = ?`;

        db.run(sql, [user.username, user.password, user.index, username], (err, row) => {
            if(err) {
                return console.log(err.message);
            }
            cb(row);
        });
    }
}

routes/auth.js
const express = require('express');
const crypto = require('crypto');
const router = express.Router();
const userdao = require('../database/userdao');
let config = require('../models/configuration');
config = new config();

router.post('/index', function (req, res, next) {
    username = req.body.username;

    if (!username) {
        res.statusMessage = 'Username missing';
        res.sendStatus(400);
    } else {
        userdao.getByUsername(username, (user) => {
            if (user) {
                // Don't actually safe the subtraction of the index. Otherwise, failed login attempts would always subtract from the index.
                // Plus, malicous users could let the index of other users run down to 0.
                res.send({ index: user.index - 1 });
            } else {
                res.sendStatus(404);
            }
        });
    }
});

router.get('/index', function (req, res, next) {
    res.send({ index: config.index });
});

router.post('/login', function (req, res, next) {
    username = req.body.username;
    password = req.body.password;

    if (!username || !password) {
        res.statusMessage = 'Username or password missing';
        res.sendStatus(400);
    } else {
        userdao.getByUsername(username, (user) => {
            if (user) {
                authenticate(user, password, (isAuthenticated) => {
                    if (isAuthenticated) {
                        req.session.loggedin = true;
                        req.session.username = user.username;
                        req.session.save();
                        res.sendStatus(200);
                    } else {
                        res.statusMessage = 'Wrong username or password';
                        res.sendStatus(401);
                    }
                });
            } else {
                res.statusMessage = 'Wrong username or password';
                res.sendStatus(401);
            }
        });
    }
});

router.get('/logout', function(req, res, next) {
    if(!req.session.loggedin) {
        res.statusMessage = 'Not logged in';
        res.sendStatus(401);
    } else {
        req.session.loggedin = undefined;
        req.session.username = undefined;
        req.session.save();
        res.sendStatus(200);
    }
});

function authenticate(user, password, cb) {
    let isAuthenticated = false;
    let hashedPassword = hash(password);
    user.index = user.index - 1;
    let maxHashingTries = config.index;

    // user.index >= 0 because hashing the password 0 times is still a legit login.
    while (!isAuthenticated && maxHashingTries > 0) {
        if (hashedPassword === user.password) {
            isAuthenticated = true;
        } else {
            maxHashingTries = maxHashingTries - 1;
            hashedPassword = hash(hashedPassword);
        }
    }

    if (isAuthenticated) {
        user.password = password;
        userdao.updateByUsername(user.username, user, () => {
            cb(isAuthenticated);
        });
    } else {
        cb(isAuthenticated);
    }
}

function hash(val) {
    return crypto.createHash('sha256').update(val).digest('hex');
}

module.exports = router;